data/blog/2025-02-01-flow-control.mdx

```diff
@@ -0,0 +1,130 @@
+---
+slug: QStash-flowControl
+title: "Parallelism and Rate Limit for Workflow And QStash"
+authors:
+    - sancar
+tags: [QStash, workflow, announcement]
+---
+## Parallelism and Rate Limits for Workflows and QStash  
+
+We are excited to announce the release of **Flow-Control**, a new feature that lets you set **Rate** and **Parallelism** limits for QStash Publish and Workflows.  
+
+This blog is divided into four sections for easier reading:  
+- [Motivation](#motivation): Why we developed Flow-Control.  
+- [How It Works](#how-it-works): Understanding rate limiting and parallelism.  
+- [How to Use It](#how-to-use-it): Practical examples of implementing Flow-Control in QStash and Workflows.  
+- [What Happened to Queues](#what-happened-to-queues): Changes from the previous queue-based system and the benefits of the new approach.  
+
+## Motivation  
+
+Our community drives our development. We’ve heard your feedback on two key points:  
+
+- **Workflows**: Our [Workflows](https://upstash.com/docs/workflow/agents/overview) product is gaining traction. However, it launched without built-in rate or parallelism limits. The existing parallelism control via Queues was unsuitable for Workflows.  
+- **Queue Limitations**: Many users relied on Queue Parallelism to prevent bursts on their endpoints. However, queues had per-plan limits due to memory/CPU allocation constraints. The new design removes these restrictions, allowing you to configure limits based on your application’s needs without worrying about plan limits.  
+
+To solve these issues, we developed Flow-Control, which works for both Workflows and QStash.  
+
+## How It Works  
+
+- **Rate Limit**: This defines the maximum number of calls per second. Calls within the same `FlowControl` key contribute to the rate count. Instead of rejecting excess calls, QStash queues them for execution in later seconds, respecting the limit.  
+- **Parallelism Limit**: This controls the number of concurrent executions. Unlike rate limiting, execution duration matters. At no time will there be more than the specified number of active calls.  
```

Comment:
What do we do if concurrency if full? Do we reject the request?

Reply:
Adding one more sentence to mention that they will wait for existing once to finish and continue.